kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,655 @@
+                        rApplyProxy.AssignLocalData(p_itr->first, p_itr->second);
+                    });
+                }
+            }
+        }
+    }
+
+    /**
+     * @brief Get the Apply Proxy object
+     *
+     * Returns the Apply proxy.
+     *
+     * @tparam TApplyFunctor
+     * @param rApplyFunctor
+     * @return ApplyProxy<TPointerDataType, TValueDataType, TApplyFunctor>
+     */
+    template <class TApplyFunctor>
+    ProxyType<TApplyFunctor> GetApplyProxy(TApplyFunctor&& rApplyFunctor)
+    {
+        return ProxyType<TApplyFunctor>(
+            std::forward<TApplyFunctor>(rApplyFunctor), mrNonLocalPointers, *this);
+    }
+
+    /**
+     * @brief Returns jobs parallel status
+     *
+     */
+    bool IsDistributed() const
+    {
+        return mrDataCommunicator.IsDistributed();
+    }
+
+    int GetMyPID() const
+    {
+        return mCurrentRank;
+    }
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    /**
+     * @brief Get keys vector from a map
+     *
+     * @tparam TKey                 Key type
+     * @tparam TArgs                Additional args
+     * @param rMap                  Map input
+     * @return std::vector<TKey>    Vector of keys
+     */
+    template<class TKey, class... TArgs>
+    static std::vector<TKey> GetKeys(const std::unordered_map<TKey, TArgs...>& rMap)
+    {
+        std::vector<TKey> keys;
+        keys.resize(rMap.size());
+
+        int local_index = 0;
+        for (const auto& r_item : rMap) {
+            keys[local_index++] = r_item.first;
+        }
+
+        return keys;
+    }
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const
+    {
+        std::stringstream buffer;
+        buffer << "GlobalPointerMapCommunicator" ;
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream& rOStream) const
+    {
+        rOStream << "GlobalPointerMapCommunicator";
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const {}
+
+    ///@}
+
+protected:
+    ///@name Protected Member Variables
+    ///@{
+
+    TGPNonLocalDataMap mrNonLocalPointers;
+    const DataCommunicator& mrDataCommunicator;
+    const int mCurrentRank;
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    void AddPointers(
+        const TGPVector& rGPVector,
+        const TValueDataType& rInitializationValue)
+    {
+        // this method is called always at the constructor, and it is not allowed to have PointerMapCommunicators
+        // created in OMP regions so, it is safer to do OMP parallelization here.
+        // this should only be called in MPI
+        IndexPartition<int>(rGPVector.size()).for_each([&](const int Index) {
+            const auto& r_gp = rGPVector(Index);
+            if (r_gp.GetRank() != mCurrentRank) {
+#pragma omp critical
```

Comment:
Actually this could be avoided if you guarantee that `rGPVector` contains unique values. Then there is no need for critical.

Probably we could even use that guarantee to check that you are not trying to initialize 2 global pointers with the same @ with different values.

Reply:
Hi @roigcarlo,

I am little bit confused, if `rGPVector` is guranteed to have unique elements, but still we need to have critical section because `mrNonLocalPointers` map will be empty at first, and it will start to add GPVector lists for remote ranks. So this `mrNonLocalPointers` key addition (remote ranks) should be done only in serial ryt? Is there a way around that?

Reply:
Ups sorry. I missread the code, you are right. (I though you were trying to prevent two accesses to `rank_local_map[r_gp]` with the same `r_gp` but the critical is for the rank.